src/nvim/api/vim.c

```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
   return keymap_array(mode, NULL);
 }
 
+/// Sets a |mapping| for the given mode.
+///
+/// Note that isn't possible to specify a target buffer when map_args contains
+/// "<buffer>"; like with |:map-<buffer>|, only the current buffer will be
+/// affected.
+///
+/// @param  map_cmd   |:map-command| for this mapping, e.g. :nmap, :unmap!,
+///                   :vnoremap, etc.
+/// @param  map_args  |:map-arguments| like "<buffer>", "<nowait>", etc.
+///                   concatenated into a single string.
+/// @param  lhs   Left-hand-side |{lhs}| of the mapping.
+/// @param  rhs   Right-hand-side |{rhs}| of the mapping.
+///
+/// @returns Zero on success, nonzero on failure.
+Integer nvim_set_keymap(String map_cmd, String map_args,
+                        String lhs, String rhs)
```

Comment:
You could add `Error *err` to the list of arguments. Then use `api_set_error()` later.